File: CPP/FrogPositionAfterTSeconds.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    double frogPosition(int n, std::vector<std::vector<int>>& edges, int t, int target) {
        std::vector<std::vector<int>> adj(n + 1);
        for (auto& e : edges) {
            adj[e[0]].push_back(e[1]);
            adj[e[1]].push_back(e[0]);
        }
        std::vector<bool> visited(n + 1, false);
        visited[1] = true;
        return dfs(adj, visited, target, t, 1, 0, 1.0);
    }

private:
    double dfs(std::vector<std::vector<int>>& adj, std::vector<bool>& visited,
               int target, int t, int node, int time, double prob) {
        std::vector<int> children;
        for (int v : adj[node]) if (!visited[v]) children.push_back(v);
        if (node == target) {
            return (time == t || children.empty()) ? prob : 0.0;
        }
        if (time == t || children.empty()) return 0.0;
        double p = prob / (double)children.size();
        for (int c : children) {
            visited[c] = true;
            double res = dfs(adj, visited, target, t, c, time + 1, p);
            if (res > 0.0) return res;
        }
        return 0.0;
    }
};

int main() {
    Solution s;
    std::vector<std::vector<int>> edges = {{1, 2}, {1, 3}, {1, 7}, {2, 4}, {2, 6}, {3, 5}};
    std::vector<std::vector<int>> empty;
    std::cout.precision(17);
    std::cout << s.frogPosition(7, edges, 2, 4) << "\n";   // 0.1666...
    std::cout << s.frogPosition(7, edges, 1, 7) << "\n";   // 0.3333...
    std::cout << s.frogPosition(7, edges, 20, 6) << "\n";  // 0.1666...
    std::cout << s.frogPosition(1, empty, 1, 1) << "\n";   // 1
    return 0;
}

File: Python/FrogPositionAfterTSeconds.py
from collections import defaultdict
from typing import List


class Solution:
    def frogPosition(self, n: int, edges: List[List[int]], t: int, target: int) -> float:
        # Tree rooted at 1. DFS while propagating probability = 1 / (# unvisited neighbors at each step).
        # Reaching target at time s: if s == t, return prob; if s < t and target is a leaf (no
        # unvisited children), frog stays put forever, also return prob; otherwise frog must move on.
        adj = defaultdict(list)
        for a, b in edges:
            adj[a].append(b)
            adj[b].append(a)

        visited = [False] * (n + 1)
        visited[1] = True

        def dfs(node: int, time: int, prob: float) -> float:
            children = [v for v in adj[node] if not visited[v]]
            if node == target:
                return prob if time == t or not children else 0.0
            if time == t or not children:
                return 0.0
            for c in children:
                visited[c] = True
                res = dfs(c, time + 1, prob / len(children))
                if res > 0:
                    return res
            return 0.0

        return dfs(1, 0, 1.0)


if __name__ == '__main__':
    s = Solution()
    edges = [[1, 2], [1, 3], [1, 7], [2, 4], [2, 6], [3, 5]]
    print(s.frogPosition(7, edges, 2, 4))  # 0.16666...
    print(s.frogPosition(7, edges, 1, 7))  # 0.33333...
    print(s.frogPosition(7, edges, 20, 6)) # 0.16666... (frog reaches 6 at t=2, stays)
    print(s.frogPosition(1, [], 1, 1))      # 1.0 (single vertex; frog never moves)
